modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/resource/v2_0/DataDefinitionResourceImpl.java

```diff
@@ -252,6 +255,29 @@ public String getDataDefinitionDataDefinitionFieldLinks(
 		).toString();
 	}
 
+	@Override
+	public Page<Permission> getDataDefinitionPermissionsPage(
+			Long dataDefinitionId, String roleNames)
+		throws Exception {
+
+		_dataDefinitionModelResourcePermission.check(
+			PermissionThreadLocal.getPermissionChecker(), dataDefinitionId,
+			ActionKeys.PERMISSIONS);
```

Comment:
@nhpatt JFYI, had to override generated permissions methods because of this :(.

we are basically delegating permission checking to each portlet that uses data engine. that is, app builder, forms and WEM will check their own permissions and not data engine's model resources permissions.

not sure if it's worth adding to the generator as this is very specific to data engine. generally each portlet, checks permissions the same way and the generated code works for them.

Reply:
Okk, yeah seems like not worth it.

Can you call the parent method? after checking the permissions?

Reply:
@nhpatt I could call the parent method if it didn't check for the PERMISSIONS permission itself.

App Builder's algorithm for checking every permission if something like that:

```
if user has AppBuilder's MANAGE permission then return true;
else if user has permission X on that resource then return true;
```

That is, if the user is an App Builder Administrator, we skip all other permissions checking, it has priority over everything else. Se for example:

https://github.com/liferay/liferay-portal/blob/master/modules/apps/app-builder/app-builder-service/src/main/java/com/liferay/app/builder/internal/data/engine/content/type/AppBuilderDataDefinitionContentType.java#L78-L92

The only we could call the generated permission method would be to remove:

```
PermissionUtil.checkPermission(
			ActionKeys.PERMISSIONS, groupLocalService, resourceName,
			dataRecordCollectionId,
			getPermissionCheckerGroupId(dataRecordCollectionId));
```

And delegate that to the child class, which I don't think it's worth it.

Reply:
Got it 👍